### PR TITLE
chore(alert): code refactoring (closes #2967)

### LINKED ISSUE #2967: BAlert - 'dismissed' event not emitted if 'dismissed-event-countdown' is listened
It seems that when listening to the `dismiss-count-down`, the `dismiss` event is not emitted.

Removing the `dismiss-count-down` allows the `dismiss` event to be emitted.

Should the `dismiss` event be emitted regardless of what is happening to the `dismiss-count-down` event?

Here is a replication fiddle - view the console for the logs.

https://jsfiddle.net/ru5ogxpv/

Additionally, the docs reference a `dismissAfterSeconds` prop, however this doesn't exist - probably supposed to reference a number in the `show` prop?

image

### PR DESCRIPTION
**Pull request by tmorehouse**
## Description of Pull Request:

Revamp of `<b-alert>` code for simpler code logic and event emitting.

Closes #2967.

To Do:
- [x] additional test scenarios
- [x] allow `show` to be a number as a string

## PR checklist:

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Enhancement to an existing feature
- [ ] ARIA accessibility
- [ ] Documentation update
- [x] Other, please describe: code refactoring

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch, **not** the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (i.e.
      `fixes #xxxx[,#xxxx]`, where "xxxx" is the issue number)
- [x] The PR should address only one issue or feature. If adding multiple features or fixing a bug
      and adding a new feature, break them into separate PRs if at all possible.
- [x] PR titles should following the
      [**Conventional Commits**](https://www.conventionalcommits.org/) naming convention (i.e.
      "fix(alert): not alerting during SSR render", "docs(badge): Updated pill examples, fix typos",
      "chore: fix typo in docs", etc). **This is very important, as the `CHANGELOG` is generated
      from these messages.**

**If new features/enhancement/fixes are added or changed:**

- [ ] Includes documentation updates (including updating the component's `package.json` for slot and
      event changes)
- [x] New/updated tests are included and passing (if required)
- [x] Existing test suites are passing
- [x] The changes have not impacted the functionality of other components or directives
- [ ] ARIA Accessibility has been taken into consideration (does it affect screen reader users or
      keyboard only users? clickable items should be in the tab index, etc)

**If adding a new feature, or changing the functionality of an existing feature, the PR's
description above includes:**

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a
      suggestion issue first and wait for approval before working on it)

### CODE REVIEW

File: src/components/alert/alert.js [hunk 1]
```diff
@@ -37,71 +36,71 @@ export default {
...
+      countDown: 0,
+      // If initially shown, we need to set these for SSR
+      localShow: Boolean(this.show),
       showClass: this.fade && this.show
     }
   },
   watch: {
     show(newVal) {
-      this.showChanged(newVal)
+      if (typeof newVal === 'number' && newVal > 0) {
```
Comment (jackmu95) [inline src/components/alert/alert.js]: `this.countDown = typeof newVal === 'number' && newVal > 0 ? newVal : 0`

File: src/components/alert/alert.js [hunk 2]
```diff
@@ -37,71 +36,71 @@ export default {
...
+      }
+      this.localShow = Boolean(newVal)
     },
-    dismissed(newVal) {
-      if (newVal) {
+    countDown(newVal) {
+      this.clearTimer()
+      this.$emit('dismiss-count-down', newVal)
+      if (this.show !== newVal) {
+        // update the v-model if needed
```
Comment (jackmu95) [inline src/components/alert/alert.js]: `// Update the v-model if needed`

File: src/components/alert/alert.js [hunk 3]
```diff
@@ -37,71 +36,67 @@ export default {
...
         this.$emit('dismissed')
       }
+      if (typeof this.show !== 'number' && this.show !== newVal) {
+        this.$emit('input', newVal)
+      }
     }
   },
+  created() {
+    this.localShow = Boolean(this.show)
+    this.countDown = typeof this.show === 'number' ? this.show : 0
```
Comment (jackmu95) [inline src/components/alert/alert.js]: What about a `parseCountDown()` method?
```js
parseCountDown(value) {
  return typeof value === 'number' && value > 0 ? value : 0
}
```

Comment (tmorehouse) [reply src/components/alert/alert.js]: Yeah that might be a good idea...

Comment (jackmu95) [reply src/components/alert/alert.js]: And maybe a `parseShow()` method too:
```js
parseShow(value) {
  return Boolean(value)
}
```

File: src/components/alert/alert.js [hunk 4]
```diff
@@ -37,71 +40,67 @@ export default {
   data() {
     return {
       countDownTimerId: null,
-      dismissed: false,
-      localShow: this.show,
+      countDown: 0,
+      // If initially shown, we need to set these for SSR
+      localShow: Boolean(this.show),
```
Comment (jackmu95) [inline src/components/alert/alert.js]: `localShow: Boolean(this.show)` => `localShow: parseShow(this.show)`

File: src/components/alert/alert.js (hunk 3 above)
Comment (tmorehouse) [reply src/components/alert/alert.js]: I might make `show` accept a string number as well (but will need to test for `show === ''` as well for boolean `true`)

File: src/components/alert/alert.js [hunk 5]
```diff
@@ -4,10 +4,33 @@ import { requestAF } from '../../utils/dom'
 
 const NAME = 'BAlert'
 
+// Convert `show` value to a number
+const parseCountDown = show => {
+  if (typeof show === 'boolean' || show = '') {
```
Comment (jackmu95) [inline src/components/alert/alert.js]: `show = ''` => `show === ''`

**Pull request merged**